### PR TITLE
Configurable Primary Orderbook

### PR DESCRIPTION
**Pull request by bh2smith**
This simple change introduces a `primary_orderbook` argument to the run time configuration in which the user can specify "Paginated" or "OnchainFiltered". 

Will eventually include "EventBased"

### CODE REVIEW

File: driver/src/main.rs [hunk 1]
```diff
@@ -102,6 +102,10 @@ struct Options {
     #[structopt(long, env = "ORDERBOOK_FILTER", default_value = "{}")]
     orderbook_filter: OrderbookFilter,
 
+    /// Primary method for orderbook retrieval ("Paginated" or "OnchainFiltered")
+    #[structopt(long, env = "PRIMARY_ORDERBOOK", default_value = "Paginated")]
+    primary_orderbook: String,
```
Comment (fleupold) [inline driver/src/main.rs:107]: Can you make this typesafe by doing something similar to what we have with `SOLVER_TYPE`. Basically creating an enum that implements FromStr. This way the logic of what cases are possible is colocated with the orderbook module and not inlined in this large main file.

**Review by nlordell — COMMENTED** [review]
Looks good! I would just recommend the change that @fleupold suggested. You can take a look at how we did it with other enums like `SchedulerKind`: https://github.com/gnosis/dex-services/blob/d787830a6940a47c92af2ceb67081787a680ebb6/driver/src/driver/scheduler.rs#L73-L107

**Review by nlordell — APPROVED** [review]
Awesome! Squeaky clean PR :rocket:

File: driver/src/orderbook/paginated_orderbook.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+use super::StableXOrderBookReading;
+use anyhow::Result;
+use ethcontract::{BlockNumber, U256};
+use std::convert::TryInto;
+use std::sync::Arc;
+
+/// Implements the StableXOrderBookReading trait by using the underlying
+/// contract in a paginated way.
+/// This avoid hitting gas limits when the total amount of orders is large.
+pub struct PaginatedStableXOrderBookReader {
```
Comment (fleupold) [inline driver/src/orderbook/paginated_orderbook.rs:14]: Next time,  refactoring like this can maybe be its own PR

**Review by fleupold — APPROVED** [review]
Nice!
